// bard-builder/dist/src/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Logger {
    constructor() {
        this.status = false;
        this.level = "info";
        if (Logger.instance)
            return Logger.instance;
        Logger.instance = this;
    }
    enable(level = "info") {
        this.status = true;
        this.level = level;
    }
    disable() {
        this.status = false;
    }
    log(type, message) {
        if (!this.status)
            return;
        switch (type) {
            case "error":
                Logger.LEVELS[this.level] <= Logger.LEVELS["error"] && console.log(`[-] Error: ${message}`);
                break;
            case "warning":
                Logger.LEVELS[this.level] <= Logger.LEVELS["debug"] && console.log(`[!] Warning: ${message}`);
                break;
            case "info":
                Logger.LEVELS[this.level] <= Logger.LEVELS["info"] && console.log(`[#] Info: ${message}`);
                break;
            case "success":
                Logger.LEVELS[this.level] <= Logger.LEVELS["info"] && console.log(`[+] Success: ${message}`);
                break;
        }
    }
}
Logger.LEVELS = {
    "error": 1,
    "debug": 2,
    "info": 3
};
const logger = new Logger();
exports.default = logger;
