// bard-builder/dist/src/utils/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Queue {
    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }
    push(value) {
        const node = { value, next: null };
        if (!this.size) {
            this.head = node;
            this.tail = node;
            node.next = this.tail;
            this.size++;
            return null;
        }
        if (this.tail)
            this.tail.next = node;
        this.tail = node;
        this.size++;
        return null;
    }
    pull() {
        if (!this.size)
            return new Error("Queue is empty");
        const node = this.head;
        if (!node)
            return new Error("Can't get any values from queue");
        if (this.size == 1) {
            this.head = null;
            this.tail = null;
            this.size--;
            return node.value;
        }
        if (!this.head)
            return new Error("Can't get any values from queue");
        this.head = this.head.next;
        this.size--;
        return node.value;
    }
    find(operator) {
        const iterator = this.items();
        let item = iterator.next();
        while (item) {
            if (operator(item.value))
                return item.value;
            item = iterator.next();
        }
        return new Error("Can't find any matches for this comparation");
    }
    *items() {
        let node = this.head;
        while (node) {
            yield node.value;
            node = node.next;
        }
    }
}
exports.default = Queue;
