// bard-builder/dist/src/utils/vow.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lock = void 0;
class Lock {
    constructor() {
        this.resolve = (value) => { };
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
        });
    }
    wait() {
        return this.promise;
    }
    release() {
        this.resolve();
    }
}
exports.Lock = Lock;
class Vow {
    handle(promise) {
        return __awaiter(this, void 0, void 0, function* () {
            return promise.then(r => r).catch(e => e);
        });
    }
    lock() {
        return new Lock();
    }
}
const vow = new Vow();
exports.default = vow;
