// bard-builder/dist/src/bot/emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmitterEvents = void 0;
var EmitterEvents;
(function (EmitterEvents) {
    EmitterEvents["ON_CREATE_SESSION"] = "on-create-session";
    EmitterEvents["ON_DELETE_SESSION"] = "on-delete-session";
    EmitterEvents["ON_EXPIRE_SESSION"] = "on-expire-session";
    EmitterEvents["ON_LOCK_SESSION"] = "on-lock-session";
    EmitterEvents["ON_UNLOCK_SESSION"] = "on-unlock-session";
    EmitterEvents["ON_REFRESH_SESSION"] = "on-refresh-session";
    EmitterEvents["ON_RECEIVE_MESSAGE"] = "on-receive-message";
    EmitterEvents["ON_SEND_MESSAGE"] = "on-send-message";
})(EmitterEvents = exports.EmitterEvents || (exports.EmitterEvents = {}));
class Emitter {
    constructor() {
        this.events = new Map();
    }
    set(event, action) {
        const item = this.events.get(event);
        if (!item)
            this.events.set(event, [action]);
        else
            item.push(action);
        return null;
    }
    get(event) {
        const item = this.events.get(event);
        if (!item)
            return new Error("Can't get any event");
        return item;
    }
    execute(event, params) {
        const actions = this.get(event);
        if (actions instanceof Error)
            return actions;
        for (const k in actions) {
            actions[k](params);
        }
        return null;
    }
}
exports.default = Emitter;
