// bard-builder/dist/src/gateway/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptyMessage = void 0;
const uuid_1 = require("uuid");
class Message {
    constructor(contact, session, origin, data, creation) {
        this.token = (0, uuid_1.v4)();
        this.contact = contact;
        this.session = session;
        this.origin = origin;
        this.data = data;
        this.creation = (creation || new Date());
    }
    validate() {
        if (!this.token.length)
            return new Error("Invalid or missing token field");
        if (!this.session.length)
            return new Error("Invalid or missing session field");
        if (!this.contact.length)
            return new Error("Invalid or missing contact field");
        if (!this.origin.length)
            return new Error("Invalid or missing origin field");
        if (this.data == null)
            return new Error("Invalid or missing data field");
        return null;
    }
}
exports.default = Message;
function createEmptyMessage() {
    return new Message("", "", "", "");
}
exports.createEmptyMessage = createEmptyMessage;
